src/goTest.ts

```diff
@@ -34,66 +45,80 @@ export function testAtCursor(goConfig: vscode.WorkspaceConfiguration, isBenchmar
-			let testFunctionName: string;
+	const getFunctions = cmd === 'benchmark' ? getBenchmarkFunctions : getTestFunctions;
 
+	editor.document.save().then(async () => {
+		try {
+			const testFunctions = await getFunctions(editor.document, null);
 			// We use functionName if it was provided as argument
 			// Otherwise find any test function containing the cursor.
-			if (args && args.functionName) {
-				testFunctionName = args.functionName;
-			} else {
-				for (let func of testFunctions) {
-					let selection = editor.selection;
-					if (selection && func.location.range.contains(selection.start)) {
-						testFunctionName = func.name;
-						break;
-					}
-				}
-			}
-
+			const testFunctionName = args && args.functionName
+				? args.functionName
+				: testFunctions.filter(func => func.location.range.contains(editor.selection.start))
+					.map(el => el.name)[0];
 			if (!testFunctionName) {
 				vscode.window.showInformationMessage('No test function found at cursor.');
 				return;
 			}
 
-			let testConfigFns = [testFunctionName];
-
-			if (!isBenchmark && extractInstanceTestName(testFunctionName)) {
-				// find test function with corresponding suite.Run
-				const testFns = findAllTestSuiteRuns(editor.document, testFunctions);
-				if (testFns) {
-					testConfigFns = testConfigFns.concat(testFns.map(t => t.name));
-				}
+			if (cmd === 'debug') {
+				await debugTestAtCursor(editor, testFunctionName, testFunctions, goConfig);
+			} else if (cmd === 'benchmark' || cmd === 'test') {
+				await runTestAtCursor(editor, testFunctionName, testFunctions, goConfig, args, cmd);
+			} else {
+				throw 'Unsupported command.';
 			}
+		} catch (err) {
+			console.error(err);
+		}
+	});
+}
 
-			const testConfig: TestConfig = {
-				goConfig: goConfig,
-				dir: path.dirname(editor.document.fileName),
-				flags: testFlags,
-				functions: testConfigFns,
-				isBenchmark: isBenchmark,
-			};
+/**
+ * Runs the test at cursor.
+ */
+async function runTestAtCursor(editor: vscode.TextEditor, testFunctionName: string, testFunctions: vscode.SymbolInformation[], goConfig: vscode.WorkspaceConfiguration, cmd: TestAtCursorCmd, args: any) {
+	const { tmpCoverPath, testFlags } = makeCoverData(goConfig, 'coverOnSingleTest', args);
 
-			// Remember this config as the last executed test.
-			lastTestConfig = testConfig;
+	const testConfigFns = cmd !== 'benchmark' && extractInstanceTestName(testFunctionName)
+		? [testFunctionName, ...findAllTestSuiteRuns(editor.document, testFunctions).map(t => t.name)]
+		: [testFunctionName];
 
-			return isModSupported(editor.document.uri).then(isMod => {
-				testConfig.isMod = isMod;
-				return goTest(testConfig).then(success => {
-					if (success && tmpCoverPath) {
-						return applyCodeCoverageToAllEditors(tmpCoverPath, testConfig.dir);
-					}
-				});
-			});
-		});
-	}).then(null, err => {
-		console.error(err);
-	});
+	const isMod = await isModSupported(editor.document.uri);
```

Comment:
here I have a question. previously the testConfig was saved without the `isMod` property. was that intentional or by accident ?

Reply:
i presume it was intended to be saved since the call to running previous test used it. if so then its fine since i just made it immutable but still is saved